server/src/main/java/io/crate/replication/logical/ShardReplicationChangesTracker.java

```diff
@@ -0,0 +1,220 @@
+ */
+
+package io.crate.replication.logical;
+
+import io.crate.common.collections.Tuple;
+import io.crate.replication.logical.action.ShardChangesAction;
+import org.apache.logging.log4j.Logger;
+import org.elasticsearch.ElasticsearchTimeoutException;
+import org.elasticsearch.client.Client;
+import org.elasticsearch.common.logging.Loggers;
+import org.elasticsearch.index.shard.IndexShard;
+import org.elasticsearch.index.shard.ShardId;
+import org.elasticsearch.threadpool.Scheduler;
+import org.elasticsearch.threadpool.ThreadPool;
+import org.elasticsearch.transport.NodeNotConnectedException;
+
+import java.io.Closeable;
+import java.io.IOException;
+import java.util.ArrayList;
+import java.util.Collections;
+import java.util.List;
+import java.util.concurrent.atomic.AtomicLong;
+import java.util.function.Consumer;
+import java.util.function.Function;
+
+import static io.crate.replication.logical.action.ShardChangesAction.TransportAction.WAIT_FOR_NEW_OPS_TIMEOUT;
+
+/**
+ * Replicates batches of {@link org.elasticsearch.index.translog.Translog.Operation}'s to the subcribers target shards.
+ *
+ * Derived from org.opensearch.replication.task.shard.ShardReplicationChangesTracker
+ */
+public class ShardReplicationChangesTracker implements Closeable {
+
+    private static final Logger LOGGER = Loggers.getLogger(ShardReplicationChangesTracker.class);
+
+    private final ShardId shardId;
+    private final LogicalReplicationSettings replicationSettings;
+    private final ThreadPool threadPool;
+    private final Function<String, Client> remoteClientResolver;
+    private final ShardTranslogSequencer sequencer;
+    private final List<Tuple<Long, Long>> missingBatches = Collections.synchronizedList(new ArrayList<>());
+    private final AtomicLong observedSeqNoAtLeader;
+    private final AtomicLong seqNoAlreadyRequested;
+    private Scheduler.ScheduledCancellable cancellable;
+
+    public ShardReplicationChangesTracker(IndexShard indexShard,
+                                          LogicalReplicationSettings replicationSettings,
+                                          ThreadPool threadPool,
+                                          Function<String, Client> remoteClientResolver,
+                                          Client client) {
+        this.shardId = indexShard.shardId();
+        this.replicationSettings = replicationSettings;
+        this.threadPool = threadPool;
+        this.remoteClientResolver = remoteClientResolver;
+        this.sequencer = new ShardTranslogSequencer(shardId, client, indexShard.getLocalCheckpoint());
+        observedSeqNoAtLeader = new AtomicLong(indexShard.getLocalCheckpoint());
+        seqNoAlreadyRequested = new AtomicLong(indexShard.getLocalCheckpoint());
+    }
+
+    public void start() {
+        LOGGER.debug("[{}] Spawning the shard changes reader", shardId);
+        var executor = threadPool.executor(ThreadPool.Names.GENERIC);
+        executor.execute(
```

Comment:
Maybe worth moving the inner block into a dedicated method and change this to `executor.execute(this::...)` ? To reduce indentation levels.